extmod/btstack/modbluetooth_btstack.c

```diff
@@ -85,6 +83,148 @@ STATIC mp_obj_bluetooth_uuid_t create_mp_uuid(uint16_t uuid16, const uint8_t *uu
+// * Raises GATT_EVENT_QUERY_COMPLETE to the supplied packet handler.
+
+// For notify/indicate/write-without-response that proceed immediately, nothing extra required.
+// For all other cases, buffer needs to be copied and protected from GC.
+// For notify/indicate:
+//  * btstack_context_callback_registration_t:
+//     * needs to be malloc'ed
+//     * needs to be protected from GC
+//     * context arg needs to point back to the callback registration so it can be freed and un-protected
+// For write-without-response
+//  * only the conn_handle is available in the callback
+//  * so we need a queue of conn_handle->(value_handle, copied buffer)
+
+// Pending operation types.
+enum {
+    // Queued for sending when possible.
+    MP_BLUETOOTH_BTSTACK_PENDING_NOTIFY, // Waiting for context callback
+    MP_BLUETOOTH_BTSTACK_PENDING_INDICATE, // Waiting for context callback
+    MP_BLUETOOTH_BTSTACK_PENDING_WRITE_NO_RESPONSE, // Waiting for conn handle
+    // Hold buffer pointer until complete.
+    MP_BLUETOOTH_BTSTACK_PENDING_WRITE, // Waiting for write done event
+};
+
+// Pending operation:
+//  - Holds a GC reference to the copied outgoing buffer.
+//  - Provides enough information for the callback handler to execute the desired operation.
+struct _mp_btstack_pending_op_t {
+    mp_btstack_pending_op_t *next; // Must be first field to match btstack_linked_item.
+
+    // See enum above.
+    uint16_t op_type;
+
+    // For all op types.
+    uint16_t conn_handle;
+
+    // For notify/indicate only.
+    // context_registration.context will point back to this struct.
+    btstack_context_callback_registration_t context_registration;
+
+    uint16_t value_handle;
+
+    // For notify/indicate/write-without-response, this is the actual buffer to send.
+    // For write-with-response, just holding onto the buffer for GC ref.
+    size_t len;
+    uint8_t buf[];
+};
+
+// Called in response to a gatts_notify/indicate being unable to complete, which then calls
+// att_server_request_to_send_notification.
+// We now have an opportunity to re-try the operation with an empty ACL buffer.
+STATIC void btstack_notify_indicate_ready_handler(void * context) {
+    MICROPY_PY_BLUETOOTH_ENTER
+    mp_btstack_pending_op_t *pending_op = (mp_btstack_pending_op_t*)context;
+    DEBUG_EVENT_printf("btstack_notify_indicate_ready_handler op_type=%d conn_handle=%d value_handle=%d len=%lu\n", pending_op->op_type, pending_op->conn_handle, pending_op->value_handle, pending_op->len);
+    if (pending_op->op_type == MP_BLUETOOTH_BTSTACK_PENDING_NOTIFY) {
+        int err = att_server_notify(pending_op->conn_handle, pending_op->value_handle, pending_op->buf, pending_op->len);
+        DEBUG_EVENT_printf("btstack_notify_indicate_ready_handler: sending notification err=%d\n", err);
+        assert(err == ERROR_CODE_SUCCESS);
+        (void)err;
+    } else if (pending_op->op_type == MP_BLUETOOTH_BTSTACK_PENDING_INDICATE) {
```

Comment:
Code size is decreased and slightly simplified if this is just an `else` and then in the body it does:
```c
assert(pending_op->op_type == MP_BLUETOOTH_BTSTACK_PENDING_INDICATE);
```
Then remove the `else` following it.